### PR TITLE
ENH: Complex step differentiation

### PR DESCRIPTION
**Pull request by Gui-FernandesBR**
## Pull request type

- [x] Code changes (bugfix, features)

## Checklist

- [x] Tests for the changes have been added (if needed)
- [ ] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [x] All tests (`pytest tests -m slow --runslow`) have passed locally
- [x] `CHANGELOG.md` has been updated (if relevant)

## Current behavior
The `Function.differentiate()` method is the only method that can be used to derivate Function objects.

## New behavior
The new, powerful, and fast `Function.differentiate_complex_step()` method is here!
This method will be used to speed up simulations in future PRs.

## Breaking change

- [x] No

## Additional information
This PR will close the #131 issue, a really old feature request of our repository.

### CODE REVIEW

File: rocketpy/mathutils/function.py [hunk 1]
```diff
@@ -2427,6 +2427,30 @@ def differentiate(self, x, dx=1e-6, order=1):
                 + self.get_value_opt(x - dx)
             ) / dx**2
 
+    def differentiate_complex_step(self, x, dx=1e-6):
+        """Differentiate a Function object at a given point using the complex
+        step method. This method can be faster than ``Function.differentiate``
```
Comment (MateusStano) [inline rocketpy/mathutils/function.py:2432]: Shouldn't `dx` be much smaller for this to be efficient?

**Review by MateusStano — COMMENTED** [review]
Really good!

Questions:

1. Are there any benefits to using the old differentiation method? Or is this one just better?
2. Have you tried plugging this in the flight class? Does it make a difference in terms of performance? If so, why not do it in this PR?
3. Could you add second-order differentiation to this?

File: rocketpy/mathutils/function.py (hunk 1 above)
Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2432]: Idk I didn't have much time to investigate it.

Copilot suggested 1e-20 but I changed to 1e-6 to make it more similar to the differentiate method.

All in all it is an user choice, I just set the standard value.
From the tests I implemented, we could see that it has to be sufficient. 

Would you still reduce the dx default value even more?

**Comment by Gui-FernandesBR** [comment]
> Really good!
> 
> Questions:
> 
> 1. Are there any benefits to using the old differentiation method? Or is this one just better?

Yes, this new method only works if the function accepts complex numbers that presents an imaginary part. Also, it is required that the function should be "differentiable in complex dimension", but this is not practical insight. Therefore, the old method should still be used sometimes.

> 2. Have you tried plugging this in the flight class? Does it make a difference in terms of performance? If so, why not do it in this PR?

Yes, it helps the flight class to go faster. The goal of this PR was to implement a complex step differentiation method, nothing more.
As I mentioned in the PR description: **"This method will be used to speed up simulations in future PRs."**

> 3. Could you add second-order differentiation to this?
- *First answer*: I have no idea.
- *After googling for 5min*: Yes, I think it is possible! See https://www.acsu.buffalo.edu/~johnc/complex_step08.pdf 

Let me try to implement it the next weekend.

File: rocketpy/mathutils/function.py (hunk 1 above)
Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2432]: Now I see the recommendation to use 1e-200, seems cool.

Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2432]: Solved here: [1356248](https://github.com/RocketPy-Team/RocketPy/pull/594/commits/13562483f646d8cb61f2105d1fa72f892aa0a37c)

**Comment by Gui-FernandesBR** [comment]
@MateusStano let's forget about the second order differentiation, it was giving some errors that I really think are not worth it to debug right now.

The tests are already covering basic scenarios and I know for a fact that the flight simulations are running correctly when using the complex step method (1st order).

I would leave the second order for a future PR.

Could you approve this one please?

**Comment by MateusStano** [comment]
> The tests are already covering basic scenarios and I know for a fact that the flight simulations are running correctly when using the complex step method (1st order).
> 
> I would leave the second order for a future PR.
> 
> Could you approve this one please?

Okay. One more thing that would be good here is using complex diff for `derivative_function`. That might make a big difference for liquids/hybrids simulations

**Review by MateusStano — APPROVED** [review]

**Pull request merged**